File: src/main/java/komu/blunt/analyzer/VariableReference.java
package komu.blunt.analyzer;

import komu.blunt.objects.Symbol;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

public final class VariableReference {

    public static final int GLOBAL_FRAME = -1;

    /** Reference to the frame of variable, 0 being local frame, or -1 being global variable */
    public final int frame;

    /** 0-based offset of variable within frame */
    public final int offset;

    public final Symbol name;

    private VariableReference(int frame, int offset, Symbol name) {
        checkArgument(frame >= 0 || frame == GLOBAL_FRAME);
        checkArgument(offset >= 0);

        this.frame = frame;
        this.offset = offset;
        this.name = checkNotNull(name);
    }

    public boolean isGlobal() {
        return frame == GLOBAL_FRAME;
    }

    public static VariableReference nested(int frame, int offset, Symbol name) {
        return new VariableReference(frame, offset, name);
    }

    public static VariableReference global(int offset, Symbol name) {
        return new VariableReference(GLOBAL_FRAME, offset, name);
    }

    @Override
    public String toString() {
        return "VariableReference [frame=" + frame + ", offset=" + offset + "]";
    }
}

File: src/main/kotlin/komu/blunt/analyzer/VariableReference.kt
package komu.blunt.analyzer

import komu.blunt.objects.Symbol

data class VariableReference private constructor(val frame: Int, val offset: Int, val name: Symbol) {

    init {
        require(frame >= GLOBAL_FRAME) { "invalid frame $frame" }
        require(offset >= 0) { "invalid offset $offset" }
    }

    companion object {
        val GLOBAL_FRAME: Int = -1

        fun nested(frame: Int, offset: Int, name: Symbol) =
            VariableReference(frame, offset, name)
    }

    val global: Boolean
        get() = frame == GLOBAL_FRAME

    override fun toString() = if (global) "(GlobalVar $offset)" else "(LocalVar $frame:$offset)"
}
